### PR TITLE
Refactor plugin framework

### PR DESCRIPTION
**Pull request by rtnpro**
Refactor plugin framework to allow plugins decide their frequency of saving data. Also, improve memoizing discovered plugins using dictionary for O(1) lookup.

### CODE REVIEW

File: statscache/schedule.py [hunk 1]
```diff
@@ -71,3 +72,40 @@ def __float__(self):
 
     def __int__(self):
         return int(float(self))
+
+
+class Frequency(object):
```
Comment (ralphbean) [inline statscache/schedule.py:103]: It's not clear to me how this Frequency class is used.
- Can you add some tests for it next to the `Schedule` tests?

Comment (ralphbean) [inline statscache/schedule.py:103]: Oh, now I see.  They're used over in fedora-infra/statscache_plugins#2.

Comment (ralphbean) [inline statscache/schedule.py:103]: Still, some test cases for it would be nice.

Comment (rtnpro) [inline statscache/schedule.py:103]: :+1:

File: statscache/schedule.py [hunk 2]
```diff
@@ -44,21 +54,38 @@ def __init__(self, second=None, minute=None, hour=None):
...
         """ Returns the datetime object representing when we'll fire next. """
-        now = datetime.datetime.utcnow()
+        now = now or datetime.datetime.utcnow()
         s, remainder = self.second.next(now, EPSILON)
         m, remainder = self.minute.next(now, remainder)
         h, remainder = self.hour.next(now, remainder)
         return datetime.datetime(now.year, now.month, now.day, h, m, s) + \
             datetime.timedelta(days=remainder)
 
+    def prev(self, now=None):
```
Comment (rtnpro) [inline statscache/schedule.py:66]: @ralphbean Are you OK with this change?

Comment (ralphbean) [inline statscache/schedule.py:66]: Ha!  yes :)

**Comment by ralphbean** [comment]
:+1: here

**Pull request merged**